### PR TITLE
Scylla shard aware

### PR DESCRIPTION
**Pull request by dahankzter**

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -331,6 +339,36 @@ func (c *Conn) startup(ctx context.Context, frameTicker chan struct{}) error {
...
+		return err
+	}
+
+	f, err := framer.parseFrame()
+	if err != nil {
+		return err
+	}
+
+	if sf, ok := f.(*supportedFrame); ok {
+		atomic.StoreInt32(&c.shardID, int32(intFromMap(sf.supported, ScyllaShardParamKey, -1)))
```
Comment (mmatczuk) [inline conn.go]: IMO atomic here is not needed as we do that before startup, with all the async we still wait for startupErr chan. Worth documenting.

File: conn.go [hunk 2]
```diff
@@ -331,6 +339,36 @@ func (c *Conn) startup(ctx context.Context, frameTicker chan struct{}) error {
...
+		s := newShardingInfoFromMap(sf.supported)
+		// atomics needed here?
+		c.host.shardingInfo = *s
```
Comment (mmatczuk) [inline conn.go]: newShardingInfoFromMap and other constructors should return value not pointer.

File: conn.go (hunk 1 above)
Comment (mmatczuk) [reply conn.go]: shardID belongs to ShardingInfo, we could have a *ShardingInfo in host and to sth like

```go
if pool.isTokenAware() && conn.host.shardingInfo != nil
```

Overall I do not like isTokenAware() && isSharded() all over the place. But it's a different story.

Comment (dahankzter) [reply conn.go]: The first atomics thing fixed. Ill think a little bit more on the other issue about dispatching for sharded functionality.

File: conn.go [hunk 3]
```diff
@@ -151,6 +151,8 @@ type Conn struct {
 	quit   chan struct{}
 
 	timeouts int64
+
+	shardID int
```
Comment (mmatczuk) [inline conn.go]: I suggest moving shardID right below host and preserve that in constructor.

**Pull request merged**